core/domain/learner_goals_services.py

```diff
@@ -0,0 +1,140 @@
+# coding: utf-8
+#
+# Copyright 2021 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Services for the learner goals feature of the learner dashboard."""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+from core.domain import user_domain
+from core.platform import models
+import feconf
+
+(user_models,) = models.Registry.import_models([models.NAMES.user])
+
+
+def get_learner_goals_from_model(learner_goals_model):
+    """Returns the learner goals domain object given the learner goals
+    model loaded from the datastore.
+
+    Args:
+        learner_goals_model: LearnerGoalsModel. The
+            learner goals model from the datastore.
+
+    Returns:
+        LearnerGoals. The learner goals domain object corresponding to the
+        given model.
+    """
+    return user_domain.LearnerGoals(
+        learner_goals_model.id,
+        learner_goals_model.topic_ids_to_learn,
+        learner_goals_model.topic_ids_to_master)
+
+
+def save_learner_goals(learner_goals):
+    """Save a learner goals domain object as an LearnerGoalsModel entity
+    in the datastore.
+
+    Args:
+        learner_goals: LearnerGoals. The learner goals domain object to
+            be saved in the datastore.
+    """
+    learner_goals_dict = learner_goals.to_dict()
+
+    learner_goals_model = (user_models.LearnerGoalsModel.get(
+        learner_goals.id, strict=False))
+    if learner_goals_model is not None:
+        learner_goals_model.populate(**learner_goals_dict)
+        learner_goals_model.update_timestamps()
+        learner_goals_model.put()
+    else:
+        learner_goals_dict['id'] = learner_goals.id
```

Comment:
This is no longer needed, right?

Reply:
I can't add the id to the `to_dict()` because it's not a property of the model. The model is keyed by the id but we don't actually have a property id. So we need this here.

Reply:
In general, to_dict represent all the property of the domain/object class and I think we have id in to_dict for other objects, so I would suggest adding `id` to to_dict!

Reply:
Yeah but `id` is not a property of the model. I checked all the other models which are keyed by the user_id (like the UserSettingsModel) and the to_dict method does not contain the id field.